### PR TITLE
[CARE-4418] Resolve NPM master version mismatch

### PR DESCRIPTION
**Pull request by kedskeds**
JIRA Ticket: [Resolve JayZ version mismatch](https://headspace.atlassian.net/browse/CARE-4418)

The version of JayZ used in Vault ([the published version on NPM](https://www.npmjs.com/package/@ginger.io/jay-z/v/0.0.14?activeTab=code)) doesn’t match JayZ’s master branch. This PR:
* copies over the files from NPM (6728909cf0ea67ff123757b5d62f4c6e38e0325a)
* applies the changes from this PR: https://github.com/HeadspaceMeditation/jay-z/pull/25

### CODE REVIEW

File: package.json
Comment (kedskeds) [inline package.json:3]: The version is being bumped twice because Vault is using JayZ `v0.0.14`, and we want to increment the version again to account for the nonce fix.

File: src/test/Jayz.test.ts
Comment (kedskeds) [inline src/test/Jayz.test.ts]: `encryptedItem` is of type `ItemWithoutEncryptedFields`, plus a metadata object. `ItemWithoutEncryptedFields` is a subset of the fields in `item`. 

However, I'm not able to call `encryptedItem`.`{key that exists on item}` without throwing an error: `Property 'pk' does not exist on type 'EncryptedJayZItem<BankAccount, keyof BankAccount>'.`

Because of that error, I casted this `as any`. I would prefer not to do that if anyone knows how to properly define the `ItemWithoutEncryptedFields` type.

File: src/main/LibsodiumEncryptor.ts
Comment (kedskeds) [inline src/main/LibsodiumEncryptor.ts:46]: This is the nonce fix. Instead of reusing the nonce in the loop, we use it once on all the fields that need to be encrypted.

File: src/main/serialization.ts
Comment (kedskeds) [inline src/main/serialization.ts]: This file was copied from the master branch. It replaces the private serialization methods on the `LibsodiumEncryptor` class.

**Comment by blimmer** [comment]
I think you'll need to manually publish from your local machine. I don't think there's a deploy step in any of the CI workflows (there probably should be).

File: package.json
Comment (gbryan) [inline package.json]: Is the downgrade here necessary?

File: package.json
Comment (gbryan) [inline package.json]: Similar here: seems like prettier was added later (on the latest commit to master that was unpublished) and it'd be nice to keep it for code formatting.

File: package.json
Comment (gbryan) [inline package.json]: From the version number, I assume this is a downgrade. Can we avoid needing to downgrade?

File: src/main/JayZ.ts
Comment (gbryan) [inline src/main/JayZ.ts]: This might be another newer change that we don't want to overwrite (though I don't think it has any functional difference). Except where a type is necessary, I think that interfaces are generally preferred now in TS.

> If you would like a heuristic, use interface until you need to use features from type.

https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#differences-between-type-aliases-and-interfaces

File: src/main/JayZ.ts
Comment (gbryan) [inline src/main/JayZ.ts]: Looks like this is a breaking change since `keyProvider` is required now. If Jay-Z's version were >= 1.0.0, we'd want to do a major version bump to indicate that breaking change. However, since it's only on 0.0.X, I think we're fine.

File: src/main/JayZ.ts
Comment (gbryan) [inline src/main/JayZ.ts:55]: Looks like some major changes in structure here vs. the previous version. We now put the encrypted fields within a metadata prop.

File: src/main/JayZ.ts
Comment (gbryan) [inline src/main/JayZ.ts]: At first glance, I don't follow how this will be backward compatible with item encryption that's been done by using the previous approach that kept the encrypted fields at the same level, rather than putting them within `__jayz__metadata`.

Maybe we don't care about backward compatibility since Vault is already using the version that we're checking in with this PR? Do we need to consider other folks using Jay-Z since it's a public repo?

Comment (gbryan) [reply src/main/JayZ.ts]: Actually, I'm wondering if we're overwriting more than we should with this PR. I see that this PR changes to setting `encryptedFieldNames` within the Jay-Z metadata, but I see in a deleted file below that the function `isLegacyJayZItem()` defines a "legacy" item as having this structure.

Should we be treating what's on `master` as the newer/preferred version and only incorporating the bug fixes from published version 0.0.14 into it?

**Review by gbryan — COMMENTED** [review]
Hey @joshc-hh, I think we could use your help understanding the changes here.

Kelsey created this PR by bringing in the published version `0.0.14` from npm. I'm concerned that we're inadvertently overwriting newer changes made on the unpublished commit on `master` (prior to the yarn berry upgrade and such that's recent).

Could you please help us understand which changes are intentional?

File: package.json
Comment (joshc-hh) [reply package.json]: Yea. I suspect what happened is the upgrade to 3.x came after the code that was extracted from the latest NPM release. 

What I'd suggest is only taking the version bump change and leaving everything else in `package.json` as it  was in `master`

File: src/main/DataKeyProvider.ts
Comment (joshc-hh) [inline src/main/DataKeyProvider.ts:1]: `DataKeyProvider` replaces `KeyProvider` and looks like (my) attempt to clarify and simplify the interfaces. 

I think the rename of `KeyProvider` => `DataKeyProvider` makes sense as there's multiple keys involved in [envelope encryption](https://docs.aws.amazon.com/wellarchitected/latest/financial-services-industry-lens/use-envelope-encryption-with-customer-master-keys.html), so clarifying that this interface is for 'data keys' is useful (imo).

Other simplifications here involve pairing down some of the fields (e.g. metadata goes away) and simplifying the `decryptDataKey` mehtod's return type from a complex type to just a `Promise<Uint8Array>`

File: src/main/Encryptor.ts
Comment (joshc-hh) [inline src/main/Encryptor.ts:6]: Note the previous `EncryptParams` and `DecryptParams` types did not ask the caller to pass in the `dataKey` used to encrypt the data. 

This yields control to the caller and allows them more control/flexibility over which data key to use for encryption.

File: src/main/Encryptor.ts
Comment (joshc-hh) [inline src/main/Encryptor.ts]: A noteable change here is that the coupling to `JayZ` types e.g. `LegacyEncryptedJayZItem` goes away, which existed in the previous code.

File: src/main/Encryptor.ts
Comment (joshc-hh) [inline src/main/Encryptor.ts:24]: Also note that the nonce has to be passed in now to decrypt.

File: src/main/Encryptor.ts
Comment (joshc-hh) [inline src/main/Encryptor.ts:22]: Also note that the caller now has the ability to specify _what_ to decrypt. Previously this wasn't a possibility as the typing assumed a JayZ encrypted item, which contains a metadata field that specifies which fields were encrypted, and thus was more rigid.

File: src/main/FixedDataKeyProvider.ts
Comment (joshc-hh) [inline src/main/FixedDataKeyProvider.ts:11]: It doesn't look like there's any material changes here outside of small changes necessary to implement the rename/tweaked `DataKeyProvider` interface.

File: package.json
Comment (joshc-hh) [inline package.json]: Given this version is already published, you'll want to open another PR later to bump the version (you can't overwrite npm packages, and generally shouldn't anyways).

File: src/main/JayZ.ts
Comment (joshc-hh) [reply src/main/JayZ.ts]: Not only that but the type name changes as well from `JayzProps` to `JayZConfig` which is also "breaking". And yea since this package is sub 1.0, breaking changes are 'allowed'. 

After this is merged you might consider just ticking the version to 1.0.

File: src/main/JayZ.ts
Comment (joshc-hh) [reply src/main/JayZ.ts]: Yea, generally most of the `type` used in this code can now safely be changed to `interface`.

File: src/main/JayZ.ts
Comment (joshc-hh) [inline src/main/JayZ.ts:19]: This was likely added to give callers a convenient way to determine if the JayZ class could be used. For context, the libsodium library is compiled to JS from C and uses WASM, so you can't use it until this `ready` promise is completed.

File: src/main/JayZ.ts
Comment (joshc-hh) [reply src/main/JayZ.ts:55]: @gbryan -- that was actually always the case. Previously the `Encryptor` interface returned a `EncryptedJayZItem` which had this metadata field in it. 

Now the `Encryptor` interface just returns the cipherText and the `JayZ` class handles appending the metadata directly. This results in a 'cleaner' interface -- i.e. encryption is just encryption and metadata 'sold separately'.

File: src/main/JayZ.ts
Comment (joshc-hh) [inline src/main/JayZ.ts:61]: The `async` tag was removed here likely because there's no need to use `await` in the body of this function.

File: src/main/JayZ.ts
Comment (joshc-hh) [reply src/main/JayZ.ts]: @gbryan 

> At first glance, I don't follow how this will be backward compatible with item encryption that's been done by using the previous approach that kept the encrypted fields at the same level, rather than putting them within __jayz__metadata.

Actually, both "legacy" _and_ "v1" use the same metadata field name. The primary difference is: 
 
1. "legacy" puts the a set of field names to decrypt into the metadata field. 
2. "v1" puts the the encrypted field values into the metadata field. 

In these changes the `Encryptor` interface was updated to handle both via its signature: `decrypt<...>(params: DecryptParams<...> | LegacyDecryptParams<...>): DecryptResult<...>`. 

Since the item to be decrypted here can either be legacy or v1, lines 80 and 81 force cast to each type. And then the call to `.decrypt` on line 87 ends up handling both cases as the parameters being passed in will either match the "legacy" signature or the "v1" signature given either `fieldsToDecrypt` or `encryptedFields` will be empty. 

This reads as me trying to be too clever/cute, it'd probably be clearer to readers if there was just a slightly longer, but more explicit check on which type of item we're dealing with.

File: src/main/JayZ.ts
Comment (joshc-hh) [inline src/main/JayZ.ts:142]: This looks to be inlined (with the goal of simplifying) from `LibsodiumKdfKeyProvider` which went away.

File: src/test/LibsodiumEncryptor.test.ts
Comment (joshc-hh) [inline src/test/LibsodiumEncryptor.test.ts:221]: This makes me feel a bit better that this nonce fix "should" be compatible with items that were encrypted prior to the fix.

File: package.json
Comment (joshc-hh) [reply package.json]: Shouldn't be. I'd recommend reverting all the package.json + .lock changes to master and running the unit tests to ensure they pass.

Comment (kedskeds) [reply package.json]: Reverted all package + lock changes to what's currently on `master`

File: src/test/Jayz.test.ts
Comment (joshc-hh) [reply src/test/Jayz.test.ts]: Hey @kedskeds, 

The reason you currently need that `as any` cast is because of the  `(keyof BankAccount)[] ` type annotation attached to the `fieldsToEncrypt` declaration on  line 15.

What that's doing is telling TypeScript that this array can contain _any_ key of `BankAccount`, and when you hit the `Omit<T, K>` bits in the encryption types that's interpreted to omit _every_ key from the type.

To Fix that, there's a few small things you'd need to do: 

Change this:

```TypeScript
const fieldsToEncrypt: (keyof BankAccount)[] = [
    "accountNumber",
    "balance",
    "routingNumber",
    "notes"
  ]
```

To this: 
```TypeScript
const fieldsToEncrypt = [
    "accountNumber",
    "balance",
    "routingNumber",
    "notes"
  ] as const
```

The `as const` will make TypeScript _narrow_ the type to the "literal" contents of the array, i.e. `readonly ["accountNumber", "balance", "routingNumber", "notes"]`, which prevents TS trying to `Omit` _every_ property on the `BankAccount` type. 

Note how that also appends a `readonly` modifier to the inferred type. So you'll need to also update `fieldsToEncrypt: K[]` declarations in the EncryptParams to `fieldsToEncrypt: readonly K[]` to stop TS complaining about readonly and mutable types being incompatible.

After that you'll get the "correct" type inferred. You'll still have a few type errors on lines 21+ since any field that's "encrypted" (e.g. `accountNumber`) won't show up as a property on the object (which is intended) and thus TS will complain -- so you'd just need to tweak that a bit.

**Comment by joshc-hh** [comment]
Thanks @kedskeds ! Overall the PR looks good to me (assuming we preserve the dependencies / lock file etc from the `master` branch). 

I've annotated several of the changes to try to provide context into "why" things changed. It's good that there's unit tests that attempt to prove that post-nonce fix we can still decrypt legacy items. But I'd recommend being careful with deploying this into a production workload, e.g. snapshot the datastore prior to deployment, test empirically in a staging environment to ensure new data can be written/read and old data can still be read w/o error etc.

Happy to help if you need anything from me.

File: src/test/Jayz.test.ts
Comment (kedskeds) [reply src/test/Jayz.test.ts]: Thank you for this excellent explanation @joshc-hh!!

File: src/main/JayZ.ts
Comment (kedskeds) [reply src/main/JayZ.ts]: Changed some of the `types` to `interface`.

File: src/main/JayZ.ts
Comment (kedskeds) [reply src/main/JayZ.ts]: I'm confused about this being a breaking change if the currently published version has `JayZConfig` with a required `keyProvider` param. wouldn't we want to preserve that over the type definition that exists on the master branch?

File: src/main/JayZ.ts
Comment (kedskeds) [reply src/main/JayZ.ts]: > This reads as me trying to be too clever/cute, it'd probably be clearer to readers if there was just a slightly longer, but more explicit check on which type of item we're dealing with.

I changed the logic so that we're not force casting the type: https://github.com/HeadspaceMeditation/jay-z/pull/39/commits/16fddff049024c0d03671574e0ee4b7f7f060dfb

Comment (gbryan) [reply src/main/JayZ.ts]: > Since the item to be decrypted here can either be legacy or v1, lines 80 and 81 force cast to each type. And then the call to .decrypt on line 87 ends up handling both cases as the parameters being passed in will either match the "legacy" signature or the "v1" signature given either fieldsToDecrypt or encryptedFields will be empty.

Gotcha, this explanation really helped me!

> I changed the logic so that we're not force casting the type: 16fddff

Thanks for doing an explicit check for the legacy structure in the `LibsodiumEncryptor`, @kedskeds! That check currently reads `if (!('encryptedFields' in params)) {`. From looking at line 92 here, it looks like the `encryptedFields` property will always be included on the object provided as the argument to `decrypt`, but it will sometimes be undefined. Do we need to check for `undefined` in `LibsodiumEncryptor` rather than checking whether the property is `in params`?

**Review by gbryan — COMMENTED** [review]
Thanks for the explanations!

@kedskeds My only remaining question is the one about the `if (!('encryptedFields' in params)) {` check.

File: src/main/JayZ.ts
Comment (kedskeds) [reply src/main/JayZ.ts]: @gbryan great catch! I added a check to see if `encryptedFields` is null / undefined.

File: src/main/JayZ.ts
Comment (gbryan) [reply src/main/JayZ.ts]: @kedskeds I think what you currently have in this PR is what we want: `JayZConfig` with a required `keyProvider`.

What I meant by "breaking" isn't really for our purposes since Vault has already been using this same interface. I was just thinking about it being a breaking change for other folks (since this is a public package) who aren't using the published-but-not-checked-in version. I don't think we need to worry about that, though. Josh's suggestion to go to version `1.0.0` once we get this merged seems reasonable.

File: src/main/JayZ.ts
Comment (gbryan) [reply src/main/JayZ.ts]: Great! If I'm reading it right, I don't think the `!('encryptedFields' in params)` will ever be true, so I think we _do_ need the `undefined` check that you added.

It'd be nice to remove that `!('encryptedFields' in params)` check as well (I know from chatting on Slack that you're already looking into it!).

**Review by gbryan — DISMISSED** [review]
Thanks, Kelsey! I can help snapshot the DynamoDB table when you're ready to start testing.

**Review by gbryan — DISMISSED** [review]

**Review by gbryan — APPROVED** [review]

**Review by gbryan — DISMISSED** [review]

**Review by gbryan — APPROVED** [review]

File: src/main/JayZ.ts
Comment (joshc-hh) [inline src/main/JayZ.ts:106]: This is a stylistic thing, but if it were me, I'd just inline this function above on line 81 as the signature of your function ends up being almost as long as the function body itself. 

One other thought here is that TypeScript's [type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) are useful for this sort of use-case where you want to check if a type is an A or a B and have TS infer narrow the type to A or B (i.e legacy or contemporary formats).

**Review by joshc-hh — APPROVED** [review]
Thanks for all the work here! 

One thing to call out for next time is that we ended up trying to refactor a fair amount of things in addition here (e.g. changing types to interfaces, cleaning up the decrypt method to be clearer etc). 

Generally it would have been a bit safer and easier to follow as a reviewer if we had done the bare minimum amount of work to merge master w/ this version, released that. And then opened up a follow up PR to make the refactoring enhancements. 

Given we've already done the work, I don't think it's worth doing that now though. And as long as the unit tests are passing you should have a fair degree of confidence that we haven't broken something a long the way. 

As a separate PR y'all should update `CODEOWNERS` so that it's pointing to you and not pointing to the tools team (this should be now managed by Terraform iirc).

**Pull request merged**
